### PR TITLE
feat: handle file content logic when creating next component version

### PR DESCRIPTION
**Pull request by Ian2012**
### Description

This PR updates `create_next_component_version` to handle raw bytes and content.pk logic to add assets. Closes https://github.com/openedx/openedx-learning/issues/241

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @Ian2012!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@axim-engineering`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: tests/openedx_learning/apps/authoring/components/test_api.py
Comment (ormsbee) [inline tests/openedx_learning/apps/authoring/components/test_api.py:445]: Please add a test that asserts that:

1. The b`raw content` is read back out at that path.
2. That the code chose the expected media type `text/plain` for it.
3. That the fallback mime type logic works when there's no way for it to guess by filename.

File: openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py:91]: Does this actually work, since you didn't modify `create_component_version_content`? There's no proper test coverage for this management command–it was a hack that I put in to get something to load test data. 😞 Sorry about that. You can add tests to this if you like, though it's not required given where the state of the code is. But please verify on your local machine that this is still working.

I think that now that you've enhanced `create_next_component_version`, we can actually do this all in one shot with the `next_version = ...` call above, and pass in the things to delete as `None` values for `content_to_replace`.

File: openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py:71]: Instead of `local_keys_to_raw_content`, can we use `local_keys_to_content_bytes`?

File: openedx_learning/apps/authoring/components/api.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/components/api.py:192]: Please update this function's docstring to reflect the changes.

File: openedx_learning/apps/authoring/components/api.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/components/api.py:203]:
```suggestion
                        component.learning_package_id,
```

This should remove two unnecessary model lookups.

File: openedx_learning/apps/authoring/components/api.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/components/api.py:195]: [request]: Please rename `content_pk` to something that indicates that it could be bytes or a primary key, e.g. `content_pk_or_bytes`.

Then later on, you can do:

```python
if isinstance(content_pk_or_bytes, bytes):
    ....
    content_pk = content.pk
else:
    content_pk = content_pk_or_bytes
```

... or something like that.

**Review by ormsbee — CHANGES_REQUESTED** [review]
Just a small handful of requests. Thank you!

File: openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py
Comment (Ian2012) [reply openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py:91]: We can actually remove the command, right? I will still try to improve the `create_component_version_content` method

File: tests/openedx_learning/apps/authoring/components/test_api.py
Comment (Ian2012) [reply tests/openedx_learning/apps/authoring/components/test_api.py]: Updated

File: openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py
Comment (Ian2012) [reply openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py]: I've updated the command implementation but I don't remember how to run it, still, I recommend deleting it if we are not going use it as the API already exists, any work would go there anyway.

Comment (ormsbee) [reply openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py]: I still find the command useful. Also, you don't need to modify `create_component_version_content`. The main place we use that is for creating the initial OLX, and we need to make our Content in a very specific way for that (custom mime-type, db-backed).

Comment (ormsbee) [reply openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py]: I think just doing `create_next_component_version` covers our 90% use case while keeping things simple overall.

Comment (Ian2012) [reply openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py]: In that case, can you quickly test it locally?

Comment (ormsbee) [reply openedx_learning/apps/authoring/components/management/commands/add_assets_to_component.py]:
```
# This was creating two images:
#
app@39772f1aa717:~/edx-platform$ python manage.py cms add_assets_to_component lib:Axim:2024-10-25 xblock.v1:html:c2c0516c-94e6-4aad-9549-51f94a62860c "static/deer.jpg:./deer.jpg" "static/hawk.jpg:./hawk.jpg"
Created v3 of xblock.v1:html:c2c0516c-94e6-4aad-9549-51f94a62860c (ad4f0345-044a-4225-a14d-d8bcf2327953):
- block.xml (5bebcc81-4374-4d1a-8104-68b2299ec48d)
- static/deer.jpg (bf16b09b-f8d7-41ee-991f-ecd099887412)
- static/hawk.jpg (2f5376aa-60c5-49ff-bb7b-5228cee9d134)

# This was deleting one:
app@39772f1aa717:~/edx-platform$ python manage.py cms add_assets_to_component lib:Axim:2024-10-25 xblock.v1:html:c2c0516c-94e6-4aad-9549-51f94a62860c "static/deer.jpg:./deer.jpg" "static/hawk.jpg:"
Created v4 of xblock.v1:html:c2c0516c-94e6-4aad-9549-51f94a62860c (f7e732e5-23dd-4771-8d7d-f03dd8fa5b96):
- block.xml (69f872ac-8928-4e01-ac81-14f11897c1d2)
- static/deer.jpg (37fed31d-31f4-4eb0-8f1d-ee8210b30f4b)

# This is replacing one:
app@39772f1aa717:~/edx-platform$ python manage.py cms add_assets_to_component lib:Axim:2024-10-25 xblock.v1:html:c2c0516c-94e6-4aad-9549-51f94a62860c "static/deer.jpg:./hawk.jpg"
Created v5 of xblock.v1:html:c2c0516c-94e6-4aad-9549-51f94a62860c (32732788-8985-4bca-b4bb-fb55fcc4586d):
- block.xml (45f08339-1597-4b7d-8cfe-374ef53bd5a0)
- static/deer.jpg (4c4d98b6-3c68-49b0-9b91-b10de772f1da)
```

All working as expected. 👍

**Review by ormsbee — APPROVED** [review]
@pomegranited: This PR looks good to me, but I'm not sure how we want to handle the version bump here. I'm also going to be out on Monday and Tuesday. If you have time, may I trouble you to please help @Ian2012 land this on Monday?

If not, it's not currently blocking work, so I can merge it when I get back on Wed.

Thank you!

**Review by pomegranited — APPROVED** [review]
:+1: Works perfectly, thank you for this PR @Ian2012 !

I left a couple of comments re documentation for your consideration.

Could you also please push a [version bump](https://github.com/openedx/openedx-learning/blob/main/openedx_learning/__init__.py)? That'll let me tag a new release after merging here.

- [x] I tested this using the updated management command:
   ```bash
   # Add file content to component
   app@928f6499df93:~/edx-platform$ ./manage.py cms add_assets_to_component lib:OpenCraft:lib2 xblock.v1:video:256f985c-89ea-49c5-9452-b52329b1bdf5 "static/avo1.jpg:/mnt/openedx-learning/avo1.jpg"
    Created v4 of xblock.v1:video:256f985c-89ea-49c5-9452-b52329b1bdf5 (eb225b73-f886-488f-8ceb-2b8108c3a85e):
   - block.xml (fe0222e8-9271-4b13-823f-c382a226ef5f)
   - static/avo1.jpg (ee6f6c25-9f1b-4044-9008-a534122ba343)
  # Remove added file from component
  app@928f6499df93:~/edx-platform$ ./manage.py cms add_assets_to_component lib:OpenCraft:lib2 xblock.v1:video:256f985c-89ea-49c5-9452-b52329b1bdf5 "static/avo1.jpg:"
  Created v5 of xblock.v1:video:256f985c-89ea-49c5-9452-b52329b1bdf5 (d8c3b2f3-07d2-4138-a73b-24147333f99e):
  - block.xml (4a66d2e7-3ab8-40a0-bc92-2f992e50b649)
  ```
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] <strike>Includes documentation</strike> N/A
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

File: openedx_learning/apps/authoring/components/api.py
Comment (ormsbee) [reply openedx_learning/apps/authoring/components/api.py:192]: FWIW, this was an oversight on my part in the review. Whenever possible, my intent was for these API functions to be atomic, so that we don't get inconsistent half-writes to the database. (Inconsistent half-writes to the file store are harder to guard against, but I feel better about that since they're idempotent inserts storing file data by hash–so "do it again" results in the same thing and hopefully the worst thing we do is write some unreferenced data.)

Comment (pomegranited) [reply openedx_learning/apps/authoring/components/api.py:192]: @ormsbee  should we do a follow-up PR to wrap this function in a transaction?

Comment (Ian2012) [reply openedx_learning/apps/authoring/components/api.py:192]: I will

Comment (ormsbee) [reply openedx_learning/apps/authoring/components/api.py:192]: 😄 Ah, that makes sense. Thank you!

**Pull request merged**
